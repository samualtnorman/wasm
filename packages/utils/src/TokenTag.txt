Keyword{
	AlignEquals,
	Block,
	Branch{,If,Table},
	Call{,Indirect},
	Data{,Drop},
	Declare,
	Drop,
	Element{,Drop},
	Else,
	End,
	Export,
	Global{,Get,Set},
	If,
	Import,
	Item,
	{Function,External}{,Reference},
	Local{,Get,Set,Tee},
	Loop,
	Memory{,Size,Grow,Fill,Copy,Initiate},
	Module,
	Mutable,
	NoOperation,
	Offset{,Equals},
	Parameter,
	Reference{Function,{,Is}Null},
	Result,
	Return,
	Select,
	Start,
	Table{,Copy,Fill,Get,Grow,Initiate,Set,Size},
	Type,
	Unreachable,
	Vector128{,Load{,{8x8,16x4,32x2}{S,Uns}igned,{8,16,32,64}{Splat,Lane},{32,64}Zero},Store{,{8,16,32,64}Lane},Const},

	Float{
		{32,64}{,
			{,Not}Equals,
			{Less,Greater}Than{,OrEquals},
			Absolute,
			Add,
			Ceiling,
			Constant,
			ConvertInteger{32,64}{S,Uns}igned,
			CopySign,
			Divide,
			Floor,
			Load,
			M{in,ax}imum,
			Multiply,
			Nearest,
			Negate,
			SquareRoot,
			Store,
			Subtract,
			Truncate
		},
		32{DemoteFloat64,ReinterpretInteger32,x4},
		64{PromoteFloat32,ReinterpretInteger64,x2}
	},

	Integer{
		{32,64}{,
			{Count{Leading,Trailing},PopCountNon}Zeros,
			{Divide,Remainder,ShiftRight,{Less,Greater}Than{,OrEquals},Truncate{,Saturate}Float{32,64}}{S,Uns}igned,
			Add,
			And,
			Constant,
			Equals{,Zero},
			Extend{8,16}Signed,
			Load{,{8,16}{S,Uns}igned},
			Multiply,
			NotEqual,
			Or,
			Rotate{Left,Right},
			ShiftLeft,
			Store{,8,16},
			Subtract,
			Xor
		},
		32{WrapInteger64,ReinterpretFloat32},
		64{{Load,ExtendInteger}32{S,Uns}igned,Store32,ReinterpretFloat64,Extend32Signed},
		8x16{,Splat,Shuffle,Swizzle},
		16x8{,Splat},
		32x4{,Splat},
		64x2{,Splat}
	}
}

Bracket{Open,Close}
Comment{Line,Block}
Error{InvalidCharacter,String{Unterminated,Invalid{Character,{,Unicode}Escape}},UnterminatedCommentBlock}
Identifier
Number
String{NonEscape,Escape{Hex,Tab,Newline,Return,Quote,Apostrophe,Backslash,Unicode},{Start,End}Quote}
UnknownKeyword
