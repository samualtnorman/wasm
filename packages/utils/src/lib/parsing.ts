import { assert } from "console"
import { MAX_LOOP_COUNT } from "../common"

export type Source = { [k: number]: unknown, length: number }
export type Index = { $: number }
export type Rule<S extends Source> = (source: S, index: Index) => boolean

export const condition = <S extends Source>(rule: Rule<S>): Rule<S> => (source, index) => {
	if (index.$ < source.length && rule(source, index)) {
		index.$++

		return true
	}

	return false
}

export const many = <S extends Source>(rule: Rule<S>): Rule<S> => (source, index) => {
	if (!rule(source, index))
		return false

	let loopsLeft = MAX_LOOP_COUNT

	while (rule(source, index))
		assert(loopsLeft--, HERE)

	return true
}

export const optional = <S extends Source>(rule: Rule<S>): Rule<S> => (source, index) => {
	rule(source, index)

	return true
}

export const negativeLookahead = <S extends Source>(rule: Rule<S>): Rule<S> => (source, index) => {
	const start = index.$

	if (rule(source, index)) {
		index.$ = start

		return false
	}

	return true
}

export const union = <S extends Source>(...rules: Rule<S>[]): Rule<S> => (source, index) => {
	for (const rule of rules) {
		if (rule(source, index))
			return true
	}

	return false
}

export const sequence = <S extends Source>(...rules: Rule<S>[]): Rule<S> => (source, index) => {
	const startIndex = index.$

	for (const rule of rules) {
		if (!rule(source, index)) {
			index.$ = startIndex

			return false
		}
	}

	return true
}

if (import.meta.vitest) {
	const { test, expect } = import.meta.vitest

	test(`negative lookahead does not move index`, () => {
		const index = { $: 0 }

		expect(negativeLookahead((_, index) => {
			index.$++

			return true
		})([], index)).toBe(false)

		expect(index.$).toBe(0)
	})
}
