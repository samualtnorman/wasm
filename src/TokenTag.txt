Keyword
Number
String
Identifier
OpenBracket
CloseBracket
Reserved
Integer32Type
Integer64Type
Float32Type
Float64Type
