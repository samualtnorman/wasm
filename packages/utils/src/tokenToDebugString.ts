import type { Token } from "./tokenise"
import { TokenTag } from "./TokenTag"

const TokenTagsToNames = Object.fromEntries(Object.entries(TokenTag).map(([ name, tag ]) => [ tag, name ]))

export const tokenToDebugString = (token: Token, code: string) =>
	`${TokenTagsToNames[token.tag]} ${JSON.stringify(code.slice(token.index, token.index + token.size))} ${token.index}`
