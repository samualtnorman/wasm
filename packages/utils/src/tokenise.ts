import { assert } from "@samual/lib/assert"
import { TokenTag } from "./TokenTag"
import { tokenToDebugString } from "./tokenToDebugString"

export type Token = { tag: TokenTag, index: number, size: number }

const NamesToKeywords = {
	Integer32: `i32`,
	Integer64: `i64`,
	Float32: `f32`,
	Float64: `f64`,
	Vector128: `v128`,
	FunctionReference: `funcref`,
	ExternalReference: `externref`,
	Function: `func`,
	External: `extern`,
	Parameter: `param`,
	Result: `result`,
	Mutable: `mut`,
	Block: `block`,
	Branch: `br`,
	BranchIf: `br_if`,
	BranchTable: `br_table`,
	Call: `call`,
	CallIndirect: `call_indirect`,
	DataDrop: `data.drop`,
	Drop: `drop`,
	ElementDrop: `elem.drop`,
	Else: `else`,
	End: `end`,
	Float32Constant: `f32.const`,
	Float32Load: `f32.load`,
	Float32Store: `f32.store`,
	Float64Constant: `f64.const`,
	Float64Load: `f64.load`,
	Float64Store: `f64.store`,
	GlobalGet: `global.get`,
	GlobalSet: `global.set`,
	If: `if`,
	Integer32Constant: `i32.const`,
	Integer32Load: `i32.load`,
	Integer32Load16Signed: `i32.load16_s`,
	Integer32Load16Unsigned: `i32.load16_u`,
	Integer32Load8Signed: `i32.load8_s`,
	Integer32Load8Unsigned: `i32.load8_u`,
	Integer32Store: `i32.store`,
	Integer32Store16: `i32.store16`,
	Integer32Store8: `i32.store8`,
	Integer64Constant: `i64.const`,
	Integer64Load: `i64.load`,
	Integer64Load16Signed: `i64.load16_s`,
	Integer64Load16Unsigned: `i64.load_16_u`,
	Integer64Load32Signed: `i64.load32_s`,
	Integer64Load32Unsigned: `i64.load32_u`,
	Integer64Load8Signed: `i64.load8_s`,
	Integer64Load8Unsigned: `i64.load8_u`,
	Integer64Store: `i64.store`,
	Integer64Store16: `i64.store16`,
	Integer64Store32: `i64.store32`,
	Integer64Store8: `i64.store8`,
	LocalGet: `local.get`,
	LocalSet: `local.set`,
	LocalTee: `local.tee`,
	Loop: `loop`,
	MemoryCopy: `memory.copy`,
	MemoryFill: `memory.fill`,
	MemoryGrow: `memory.grow`,
	MemoryInitiate: `memory.init`,
	MemorySize: `memory.size`,
	NoOperation: `nop`,
	ReferenceFunction: `ref.func`,
	ReferenceIsNull: `ref.is_null`,
	ReferenceNull: `ref.null`,
	Return: `return`,
	Select: `select`,
	TableCopy: `table.copy`,
	TableFill: `table.fill`,
	TableGet: `table.get`,
	TableGrow: `table.grow`,
	TableInitiate: `table.init`,
	TableSet: `table.set`,
	TableSize: `table.size`,
	Unreachable: `unreachable`,
	Module: `module`,
	Data: `data`,
	Declare: `declare`,
	Element: `elem`,
	Export: `export`,
	Global: `global`,
	Import: `import`,
	Item: `item`,
	Local: `local`,
	Memory: `memory`,
	Offset: `offset`,
	Start: `start`,
	Table: `table`,
	Type: `type`,
	Float32Absolute: `f32.abs`,
	Float32Add: `f32.add`,
	Float32Ceiling: `f32.ceil`,
	Float32ConvertInteger32Signed: `f32.convert_i32_s`,
	Float32ConvertInteger32Unsigned: `f32.convert_i32_u`,
	Float32ConvertInteger64Signed: `f32.convert_i64_s`,
	Float32ConvertInteger64Unsigned: `f32.convert_i64_u`,
	Float32CopySign: `f32.copy_s`,
	Float32DemoteFloat64: `f32.demote_f64`,
	Float32Divide: `f32.div`,
	Float32Equals: `f32.eq`,
	Float32Floor: `f32.floor`,
	Float32GreaterThan: `f32.gt`,
	Float32GreaterThanOrEquals: `f32.ge`,
	Float32LessThan: `f32.lt`,
	Float32LessThanOrEquals: `f32.le`,
	Float32Maximum: `f32.max`,
	Float32Minimum: `f32.min`,
	Float32Multiply: `f32.mul`,
	Float32Nearest: `f32.nearest`,
	Float32Negate: `f32.neg`,
	Float32NotEquals: `f32.ne`,
	Float32ReinterpretInteger32: `f32.reinterpret_i32`,
	Float32SquareRoot: `f32.sqrt`,
	Float32Subtract: `f32.sub`,
	Float32Truncate: `f32.trunc`,
	Float64Absolute: `f64.abs`,
	Float64Add: `f64.add`,
	Float64Ceiling: `f64.ceil`,
	Float64ConvertInteger32Signed: `f64.convert_i32_s`,
	Float64ConvertInteger32Unsigned: `f64.convert_i32_u`,
	Float64ConvertInteger64Signed: `f64.convert_i64_s`,
	Float64ConvertInteger64Unsigned: `f64.convert_i64_u`,
	Float64CopySign: `f64.copy_s`,
	Float64Divide: `f64.div`,
	Float64Equals: `f64.eq`,
	Float64Floor: `f64.floor`,
	Float64GreaterThan: `f64.gt`,
	Float64GreaterThanOrEquals: `f64.ge`,
	Float64LessThan: `f64.lt`,
	Float64LessThanOrEquals: `f64.le`,
	Float64Maximum: `f64.max`,
	Float64Minimum: `f64.min`,
	Float64Multiply: `f64.mul`,
	Float64Nearest: `f64.nearest`,
	Float64Negate: `f64.neg`,
	Float64NotEquals: `f64.ne`,
	Float64PromoteFloat32: `f64.promote_f32`,
	Float64ReinterpretInteger64: `f64.reinterpret_i64`,
	Float64SquareRoot: `f64.sqrt`,
	Float64Subtract: `f64.sub`,
	Float64Truncate: `f64.trunc`,
	Integer32Add: `i32.add`,
	Integer32And: `i32.and`,
	Integer32CountLeadingZeros: `i32.clz`,
	Integer32CountTrailingZeros: `i32.ctz`,
	Integer32DivideSigned: `i32.div_s`,
	Integer32DivideUnsigned: `i32.div_u`,
	Integer32Equals: `i32.eq`,
	Integer32EqualsZero: `i32.eqz`,
	Integer32Extend16Signed: `i32.extend_16_s`,
	Integer32Extend8Signed: `i32.extend_8_s`,
	Integer32GreaterThanOrEqualsSigned: `i32.ge_s`,
	Integer32GreaterThanOrEqualsUnsigned: `i32.ge_u`,
	Integer32GreaterThanSigned: `i32.gt_s`,
	Integer32GreaterThanUnsigned: `i32.gt_u`,
	Integer32LessThanOrEqualsSigned: `i32.le_s`,
	Integer32LessThanOrEqualsUnsigned: `i32.le_u`,
	Integer32LessThanSigned: `i32.lt_s`,
	Integer32LessThanUnsigned: `i32.lt_u`,
	Integer32Multiply: `i32.mul`,
	Integer32NotEqual: `i32.ne`,
	Integer32Or: `i32.or`,
	Integer32PopCountNonZeros: `i32.popcnt`,
	Integer32ReinterpretFloat32: `i32.reinterpret_f32`,
	Integer32RemainderSigned: `i32.rem_s`,
	Integer32RemainderUnsigned: `i32.rem_u`,
	Integer32RotateLeft: `i32.rotl`,
	Integer32RotateRight: `i32.rotr`,
	Integer32ShiftLeft: `i32.shl`,
	Integer32ShiftRightSigned: `i32.shr_s`,
	Integer32ShiftRightUnsigned: `i32.shr_u`,
	Integer32Subtract: `i32.sub`,
	Integer32TruncateFloat32Signed: `i32.trunc_f32_s`,
	Integer32TruncateFloat32Unsigned: `i32.trunc_f32_u`,
	Integer32TruncateFloat64Signed: `i32.trunc_f64_s`,
	Integer32TruncateFloat64Unsigned: `i32.trunc_f64_u`,
	Integer32TruncateSaturateFloat32Signed: `i32.trunc_sat_f32_s`,
	Integer32TruncateSaturateFloat32Unsigned: `i32.trunc_sat_f32_u`,
	Integer32TruncateSaturateFloat64Signed: `i32.trunc_sat_f64_s`,
	Integer32TruncateSaturateFloat64Unsigned: `i32.trunc_sat_f64_u`,
	Integer32WrapInteger64: `i32.wrap_i64`,
	Integer32Xor: `i32.xor`,
	Integer64Add: `i64.add`,
	Integer64And: `i64.and`,
	Integer64CountLeadingZeros: `i64.clz`,
	Integer64CountTrailingZeros: `i64.ctz`,
	Integer64DivideSigned: `i64.div_s`,
	Integer64DivideUnsigned: `i64.div_u`,
	Integer64Equals: `i64.eq`,
	Integer64EqualsZero: `i64.eqz`,
	Integer64Extend16Signed: `i64.extend_16_s`,
	Integer64Extend8Signed: `i64.extend_8_s`,
	Integer64GreaterThanOrEqualsSigned: `i64.ge_s`,
	Integer64GreaterThanOrEqualsUnsigned: `i64.ge_u`,
	Integer64GreaterThanSigned: `i64.gt_s`,
	Integer64GreaterThanUnsigned: `i64.gt_u`,
	Integer64LessThanOrEqualsSigned: `i64.le_s`,
	Integer64LessThanOrEqualsUnsigned: `i64.le_u`,
	Integer64LessThanSigned: `i64.lt_s`,
	Integer64LessThanUnsigned: `i64.lt_u`,
	Integer64Multiply: `i64.mul`,
	Integer64NotEqual: `i64.ne`,
	Integer64Or: `i64.or`,
	Integer64PopCountNonZeros: `i64.popcnt`,
	Integer64ReinterpretFloat64: `i64.reinterpret_f64`,
	Integer64RemainderSigned: `i64.rem_s`,
	Integer64RemainderUnsigned: `i64.rem_u`,
	Integer64RotateLeft: `i64.rotl`,
	Integer64RotateRight: `i64.rotr`,
	Integer64ShiftLeft: `i64.shl`,
	Integer64ShiftRightSigned: `i64.shr_s`,
	Integer64ShiftRightUnsigned: `i64.shr_u`,
	Integer64Subtract: `i64.sub`,
	Integer64TruncateFloat64Signed: `i64.trunc_f64_s`,
	Integer64TruncateFloat64Unsigned: `i64.trunc_f64_u`,
	Integer64TruncateSaturateFloat64Signed: `i64.trunc_sat_f64_s`,
	Integer64TruncateSaturateFloat64Unsigned: `i64.trunc_sat_f64_u`,
	Integer64Xor: `i64.xor`,
	Integer64Extend32Signed: `i64.extend_32_s`,
	Integer64ExtendInteger32Signed: `i64.extend_i32_s`,
	Integer64ExtendInteger32Unsigned: `i64.extend_i32_u`,
	Integer64TruncateFloat32Signed: `i64.trunc_f32_s`,
	Integer64TruncateFloat32Unsigned: `i64.trunc_f32_u`,
	Integer64TruncateSaturateFloat32Signed: `i64.trunc_sat_f32_s`,
	Integer64TruncateSaturateFloat32Unsigned: `i64.trunc_sat_f32_u`
} satisfies {
	[K in Exclude<
		keyof typeof TokenTag,
		`Keyword` | `Number` | `String` | `Identifier` | `OpenBracket` | `CloseBracket` | `LineComment` |
		`BlockComment` | `OffsetEquals` | `AlignEquals` | `Error`
	>]: string
}

export function* tokenise(code: string): Generator<Token, void, void> {
	let index = 0

	const optional = (rule: () => boolean) => () => {
		rule()

		return true
	}

	const range = (first: string, last: string) => () => {
		assert(first < last, HERE)

		if (code[index]! >= first && code[index]! <= last) {
			index++

			return true
		}

		return false
	}

	const oneOf = (search: string) => () => {
		if (index < code.length && search.includes(code[ index ]!)) {
			index++

			return true
		}

		return false
	}

	const union = (...rules: (() => boolean)[]) => () => {
		for (const rule of rules) {
			if (rule())
				return true
		}

		return false
	}

	const condition = (checker: () => boolean) => () => {
		if (index < code.length && checker()) {
			index++

			return true
		}

		return false
	}

	const terminal = (search: string) => () => {
		if (code.startsWith(search, index)) {
			index += search.length

			return true
		}

		return false
	}

	const sequence = (...rules: (() => boolean)[]) => () => {
		const startIndex = index

		for (const rule of rules) {
			if (!rule()) {
				index = startIndex

				return false
			}
		}

		return true
	}

	const many = (rule: () => boolean) => () => {
		if (!rule())
			return false

		while (rule());

		return true
	}

	const negativeLookahead = (rule: () => boolean) => () => {
		const start = index

		if (rule()) {
			index = start

			return false
		}

		return true
	}

	const Newline = union(oneOf(`\n\r`), terminal(`\r\n`))
	const Format = union(Newline, terminal(`\t`))

	const IdentifierCharacter =
		union(range(`0`, `9`), range(`A`, `Z`), range(`a`, `z`), oneOf(`!#$%&'*+-./:<=>?@\\^_\`|~`))

	const Keyword = sequence(range(`a`, `z`), optional(many(IdentifierCharacter)))
	const HexDigit = union(range(`0`, `9`), range(`A`, `F`), range(`a`, `f`))
	const HexNumber = sequence(HexDigit, optional(many(sequence(optional(terminal(`_`)), HexDigit))))

	const StringCharacter = union(
		condition(() => code[index]! >= `\x20` && code[index]! != `\x7F` && code[index] != `"` && code[index] != `\\`),
		terminal(`\\t`),
		terminal(`\\n`),
		terminal(`\\r`),
		terminal(`\\"`),
		terminal(`\\'`),
		terminal(`\\\\`),
		sequence(terminal(`\\u{`), HexNumber, terminal(`}`))
	)

	const StringElement = union(StringCharacter, sequence(terminal(`\\`), HexDigit, HexDigit))
	const String = sequence(terminal(`"`), optional(many(StringElement)), terminal(`"`))
	const Identifier = sequence(terminal(`$`), many(IdentifierCharacter))
	const Number = sequence(range(`0`, `9`), optional(many(sequence(optional(terminal(`_`)), range(`0`, `9`)))))
	const LineCharacter = condition(() => code[index] != `\n` && code[index] != `\r`)

	const LineComment =
		sequence(terminal(`;;`), optional(many(LineCharacter)), union(Newline, () => index == code.length))

	const BlockCharacter = union(
		condition(() => code[index] != `;` && code[index] != `(`),
		condition(() => code[index] == `;` && code[index + 1] != `)`),
		condition(() => code[index] == `(` && code[index + 1] != `;`),
		() => BlockComment()
	)

	const BlockComment = sequence(terminal(`(;`), optional(many(BlockCharacter)), terminal(`;)`))
	const Space = many(union(terminal(` `), Format))
	const Sign = optional(union(terminal(`+`), terminal(`-`)))
	const Fraction = sequence(range(`0`, `9`), optional(many(sequence(optional(terminal(`_`)), range(`0`, `9`)))))
	const HexFraction = sequence(HexDigit, optional(many(sequence(optional(terminal(`_`)), HexDigit))))

	const DecimalFloat = sequence(
		Number,
		union(sequence(terminal(`.`), Fraction), optional(terminal(`.`))),
		optional(sequence(oneOf(`Ee`), Sign, Number))
	)

	const HexFloat = sequence(
		terminal(`0x`),
		HexNumber,
		union(sequence(terminal(`.`), HexFraction), optional(terminal(`.`))),
		optional(sequence(oneOf(`Pp`), Sign, HexNumber))
	)

	const FloatMag =
		union(HexFloat, DecimalFloat, terminal(`inf`), sequence(terminal(`nan:0x`), HexNumber), terminal(`nan`))

	const OffsetEquals = terminal(`offset=`)
	const AlignEquals = terminal(`align=`)
	const Float = sequence(Sign, FloatMag)
	const OpenBracket = terminal(`(`)
	const CloseBracket = terminal(`)`)

	const tokenFunctions: Record<Exclude<keyof typeof TokenTag, "Error">, () => boolean> = {
		...Object.fromEntries(Object.entries(NamesToKeywords)
			.map(([ name, keyword ]) => [ name, sequence(terminal(keyword), negativeLookahead(IdentifierCharacter)) ])
		) as Record<keyof typeof NamesToKeywords, () => boolean>,
		AlignEquals, OffsetEquals, LineComment, BlockComment, Number: Float, Keyword, String, Identifier, OpenBracket,
		CloseBracket
	}

	let errorIndex: number | undefined
	const Token = (tag: TokenTag, startIndex: number) => ({ tag, index: startIndex, size: index - startIndex })

	while_: while (index < code.length) {
		if (Space()) {
			if (errorIndex != undefined) {
				yield Token(TokenTag.Error, errorIndex)
				errorIndex = undefined
			}

			continue
		}

		const startIndex = index

		for (const name in tokenFunctions) {
			if (tokenFunctions[name as keyof typeof tokenFunctions]()) {
				if (errorIndex != undefined) {
					yield Token(TokenTag.Error, errorIndex)
					errorIndex = undefined
				}

				yield Token(TokenTag[name as keyof typeof tokenFunctions], startIndex)
				continue while_
			}
		}

		if (errorIndex == undefined)
			errorIndex = index

		index++
	}

	if (errorIndex != undefined)
		yield Token(TokenTag.Error, errorIndex)
}

if (import.meta.vitest) {
	const { test, expect } = import.meta.vitest

	expect.addSnapshotSerializer({
		serialize: (code: string) => [ ...tokenise(code) ].map(token => tokenToDebugString(token, code)).join("\n"),
		test: _ => true
	})

	test(`tokenise wasi hello world`, () => expect(`
		(import "wasi_unstable" "fd_write" (func $fd_write (param i32 i32 i32 i32) (result i32)))

		(memory 1)
		(export "memory" (memory 0))

		;; io vector
		(data (i32.const 0) "\\08") ;; pointer to string
		(data (i32.const 4) "\\0e") ;; length of string
		(data (i32.const 8) "Hello, World!\\n")

		(global $stdout i32 (i32.const 1))

		(func (export "_start")
			(drop
				(call $fd_write
					(global.get $stdout)
					(i32.const 0) ;; pointer to io vector
					(i32.const 1) ;; length of io vector
					(i32.const 20) ;; where to store number of bytes written
				)
			)
		)
	`).toMatchSnapshot())

	test(`line comment`, () => expect(`;;`).toMatchSnapshot())
	test(`nested comment`, () => expect(`(; a (; b ;) c ;)`).toMatchSnapshot())
	test(`signed integer`, () => expect(`+123`).toMatchSnapshot())
	test(`number`, () => expect(`3.`).toMatchSnapshot())
	test(`number with fraction`, () => expect(`3.5`).toMatchSnapshot())
	test(`number with exponent`, () => expect(`1e2`).toMatchSnapshot())
	test(`hex`, () => expect(`0x1`).toMatchSnapshot())
	test(`hex exponent`, () => expect(`0x1p1`).toMatchSnapshot())
	test(`funci32`, () => expect(`funci32`).toMatchSnapshot())
	test(`offset=0`, () => expect(`offset=0`).toMatchSnapshot())
	test(`align=0`, () => expect(`align=0`).toMatchSnapshot())
	test(`error token`, () => expect(`,`).toMatchSnapshot())
	test(`collapse error tokens`, () => expect(`,,`).toMatchSnapshot())
	test(`seperate error tokens around whitespace`, () => expect(`, ,`).toMatchSnapshot())
}
