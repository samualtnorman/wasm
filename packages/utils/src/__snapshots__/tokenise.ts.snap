// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`align=0 1`] = `
AlignEquals "align=" 0
Number "0" 6
`;

exports[`collapse error tokens 1`] = `Error ",," 0`;

exports[`error then keyword 1`] = `
Error "," 0
Integer32 "i32" 1
`;

exports[`error token 1`] = `Error "," 0`;

exports[`funci32 1`] = `Keyword "funci32" 0`;

exports[`hex 1`] = `Number "0x1" 0`;

exports[`hex exponent 1`] = `Number "0x1p1" 0`;

exports[`line comment 1`] = `LineComment ";;" 0`;

exports[`nested comment 1`] = `BlockComment "(; a (; b ;) c ;)" 0`;

exports[`number 1`] = `Number "3." 0`;

exports[`number with exponent 1`] = `Number "1e2" 0`;

exports[`number with fraction 1`] = `Number "3.5" 0`;

exports[`offset=0 1`] = `
OffsetEquals "offset=" 0
Number "0" 7
`;

exports[`seperate error tokens around whitespace 1`] = `
Error "," 0
Error "," 2
`;

exports[`signed integer 1`] = `Number "+123" 0`;

exports[`tokenise wasi hello world 1`] = `
OpenBracket "(" 3
Import "import" 4
String "\\"wasi_unstable\\"" 11
String "\\"fd_write\\"" 27
OpenBracket "(" 38
Function "func" 39
Identifier "$fd_write" 44
OpenBracket "(" 54
Parameter "param" 55
Integer32 "i32" 61
Integer32 "i32" 65
Integer32 "i32" 69
Integer32 "i32" 73
CloseBracket ")" 76
OpenBracket "(" 78
Result "result" 79
Integer32 "i32" 86
CloseBracket ")" 89
CloseBracket ")" 90
CloseBracket ")" 91
OpenBracket "(" 96
Memory "memory" 97
Number "1" 104
CloseBracket ")" 105
OpenBracket "(" 109
Export "export" 110
String "\\"memory\\"" 117
OpenBracket "(" 126
Memory "memory" 127
Number "0" 134
CloseBracket ")" 135
CloseBracket ")" 136
LineComment ";; io vector\\n" 141
OpenBracket "(" 156
Data "data" 157
OpenBracket "(" 162
Integer32Constant "i32.const" 163
Number "0" 173
CloseBracket ")" 174
String "\\"\\\\08\\"" 176
CloseBracket ")" 181
LineComment ";; pointer to string\\n" 183
OpenBracket "(" 206
Data "data" 207
OpenBracket "(" 212
Integer32Constant "i32.const" 213
Number "4" 223
CloseBracket ")" 224
String "\\"\\\\0e\\"" 226
CloseBracket ")" 231
LineComment ";; length of string\\n" 233
OpenBracket "(" 255
Data "data" 256
OpenBracket "(" 261
Integer32Constant "i32.const" 262
Number "8" 272
CloseBracket ")" 273
String "\\"Hello, World!\\\\n\\"" 275
CloseBracket ")" 292
OpenBracket "(" 297
Global "global" 298
Identifier "$stdout" 305
Integer32 "i32" 313
OpenBracket "(" 317
Integer32Constant "i32.const" 318
Number "1" 328
CloseBracket ")" 329
CloseBracket ")" 330
OpenBracket "(" 335
Function "func" 336
OpenBracket "(" 341
Export "export" 342
String "\\"_start\\"" 349
CloseBracket ")" 357
OpenBracket "(" 362
Drop "drop" 363
OpenBracket "(" 372
Call "call" 373
Identifier "$fd_write" 378
OpenBracket "(" 393
GlobalGet "global.get" 394
Identifier "$stdout" 405
CloseBracket ")" 412
OpenBracket "(" 419
Integer32Constant "i32.const" 420
Number "0" 430
CloseBracket ")" 431
LineComment ";; pointer to io vector\\n" 433
OpenBracket "(" 462
Integer32Constant "i32.const" 463
Number "1" 473
CloseBracket ")" 474
LineComment ";; length of io vector\\n" 476
OpenBracket "(" 504
Integer32Constant "i32.const" 505
Number "20" 515
CloseBracket ")" 517
LineComment ";; where to store number of bytes written\\n" 519
CloseBracket ")" 565
CloseBracket ")" 570
CloseBracket ")" 574
`;
